### PR TITLE
lab 04 updated, first commit

### LINKED ISSUE #448: Lab 04 Numeronym
/* Issue to be closed when pull request approved and merged */

Create an executable go program in directory 04_numeronym/marchitect
Write a function that returns a slice of numeronyms for its input strings

### PR DESCRIPTION
**Pull request by Marchitect**
Fixes #448 

Review of colleague's PR #424 

#### Changes proposed in this PR:

- added function numeronyms() and corresponding test cases

### CODE REVIEW

**Comment by patrickmarabeas** [comment]
Looks good to me; nice accounting for special characters. The only suggestion I have is your  commit message read more like the PR message - meaningful and descriptive about what and why.

**Comment by Marchitect** [comment]
> Looks good to me; nice accounting for special characters. The only suggestion I have is your commit message read more like the PR message - meaningful and descriptive about what and why.

Thank you :) I just learnt today to use the git rebase command to make changes to commit messages, will make adjustment.

File: 04_numeronym/marchitect/main.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"strconv"
+)
+
+var out io.Writer = os.Stdout
+
+func numeronyms(vals ...string) []string {
+	result := make([]string, 0)
+	for _, val := range vals {
+		numeronym := ""
+		length := len([]rune(val))
```
Comment (nicko-lee) [inline 04_numeronym/marchitect/main.go:16]: It's good you took into account Unicode characters. Initially when I wrote my function it worked for normal english alphabet characters. But once I put something like an emoji it gave me weird results until I converted them to a rune as well

**Review by nicko-lee — APPROVED** [review]

File: 04_numeronym/marchitect/main.go (hunk 1 above)
Comment (juliaogris) [inline 04_numeronym/marchitect/main.go:16]: Hold on to []rune conversion as you use it again down below, e.g.:
```go
runes := []rune(val)
```
(conversion is O(N), N being the number of chars).

File: 04_numeronym/marchitect/main.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+)
+
+var out io.Writer = os.Stdout
+
+func numeronyms(vals ...string) []string {
+	result := make([]string, 0)
```
Comment (juliaogris) [inline 04_numeronym/marchitect/main.go:13]: Add capacity.

File: 04_numeronym/marchitect/main.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	for _, val := range vals {
```
Comment (juliaogris) [inline 04_numeronym/marchitect/main.go:14]: IMO it's nicer to have two functions: `numeronyms` and `numeronym`:
```go
func numeronyms(vals ...string) []string {
	result := make([]string, len(vals))
	for i, val := range vals {
                result[i] =  numeronym(val)
        }
        return result
}

func numeronyms(s string) string {
    // ...
}
```

File: 04_numeronym/marchitect/main.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		if length <= 3 {
+			numeronym = val
+		} else {
+			// convert rune slice to string using typecast string()
+			// why use typecast []rune? because special character can take more than 1 byte space
+			numeronym = string([]rune(val)[0]) + strconv.Itoa(length-2) + string([]rune(val)[length-1])
+		}
+		result = append(result, numeronym)
+	}
+
```
Comment (juliaogris) [inline 04_numeronym/marchitect/main.go:26]: Unnecessary blank line.
No need for blank lines at beginning or end of block.
Use blank lines between type, method and function definitions and within a block for logical separation.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please [fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) with `git rebase -i HASH`

**Pull request closed without merging**